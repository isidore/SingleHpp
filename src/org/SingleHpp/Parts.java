package org.SingleHpp;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Parts implements Comparable<Parts> {
	public ArrayList<String> includes = new ArrayList<>();
	public String code;
	public String fileName;

	@Override
	public int compareTo(Parts that) {
		if (this.dependsOn(that)) {
			return 1;
		}
		if (that.dependsOn(this)) {
			return -1;
		}
		return 0;
	}

	public boolean dependsOn(Parts that) {
		List<String> includesNames = Arrays
				.asList(this.includes.stream().map(i -> getFileFromInclude(i)).toArray(String[]::new));
		return includesNames.contains(that.fileName);
	}

	@Override
	public String toString() {
		return "Parts [" + fileName + "]";
	}

	public static String getFileFromInclude(String include) {
		StringBuffer b = new StringBuffer();
		SingleHpp.replaceAll(include, "(<[^>]+>)|(\".*?\")", m -> {
			b.append(SingleHpp.getAfterLast("/", SingleHpp.substring(m.group(), 1, -1)));
			return "";
		});
		return b.toString();
	}
}
