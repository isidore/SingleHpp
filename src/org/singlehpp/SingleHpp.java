package org.singlehpp;

import java.io.File;
import java.io.FileFilter;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.UUID;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Stream;

import org.lambda.functions.Function1;

import com.spun.util.io.FileUtils;
import com.spun.util.logger.SimpleLogger;

public class SingleHpp {

	public static List<File> getAllFiles(String directory) {
		FileFilter filter = p -> p.getName().endsWith(".h") || p.getName().endsWith(".cpp");
		File[] list = FileUtils.getRecursiveFileList(new File(directory), filter);
		Arrays.sort(list);
		return Arrays.asList(list);
	}

	public static String createTextFor(String directory) throws IOException {

		List<File> files = getAllFiles(directory);
		List<Parts> parts = getParts(files);

		return combineParts(parts, files);
	}

	public static List<Parts> getParts(List<File> files) throws IOException {
		List<Parts> parts = new ArrayList<>();
		for (File file : files) {
			parts.add(processFile(file));
		}
		return sortPartsByDependencies(parts);
	}

	private static List<Parts> sortPartsByDependencies(List<Parts> parts) {
		for (int i = 0; i < parts.size(); i++) {
			Parts lowest = parts.get(i);
			for (int j = i + 1; j < parts.size(); j++) {
				Parts next = parts.get(j);
				if (lowest.dependsOn(next)) {
					parts.remove(j);
					parts.add(0, next);
					i = -1;
					break;
				}
			}
		}
		return parts;
	}

	private static String combineParts(List<Parts> parts, List<File> files) {
		StringBuffer text = new StringBuffer();
		List<String> headers = reduceHeaders(parts, files);
		for (String header : headers) {
			text.append(header);
		}
		for (Parts p : parts) {
			text.append(" // ******************** From: " + p.fileName + "\n");
			text.append(p.code + "\n");
		}
		return text.toString();
	}

	private static List<String> reduceHeaders(List<Parts> parts, List<File> files) {
		ArrayList<String> headers = new ArrayList<>();
		String[] flatMap = parts.stream().flatMap(p -> p.includes.stream()).toArray(String[]::new);
		String[] paths = files.stream().map(f -> f.getAbsolutePath()).toArray(String[]::new);
		for (String header : flatMap) {
			if (!headers.contains(header)
					&& !Stream.of(paths).anyMatch(p -> p.endsWith(Parts.getFileFromInclude(header)))) {
				headers.add(header);
			}
		}
		return headers;
	}

	public static String getAfterLast(String match, String fromString) {
		int lastIndexOf = fromString.lastIndexOf(match);
		if (0 < lastIndexOf) {
			return fromString.substring(lastIndexOf + 1);
		} else {
			return fromString;
		}

	}

	public static String substring(String string, int start, int end) {
		end = end < 0 ? string.length() + end : end;
		if (0 <= start && 0 <= end && end <= string.length()) {
			return string.substring(start, end);
		} else {
			return "";
		}
	}

	public static String removeComments(String readFile) {
		return readFile.replaceAll("(?:/\\*(?:[^*]|(?:\\*+[^*/]))*\\*+/)|(?://.*)", "");
	}

	public static String replaceAll(String text, String regex, Function1<Matcher, String> replacer) {
		Pattern pattern = Pattern.compile(regex);
		Matcher matcher = pattern.matcher(text);
		StringBuffer result = new StringBuffer();
		while (matcher.find()) {
			matcher.appendReplacement(result, replacer.call(matcher));
		}
		matcher.appendTail(result);
		return result.toString();
	}

	public static Parts processFile(File file) throws IOException {
		String fullText = FileUtils.readFile(file);
		HashMap<String, String> unalterable = new HashMap<>();
		String masked = handledUnalterable(fullText, unalterable);
		String code = removeComments(masked);
		final Parts p = new Parts();

		String maskedMinusIncludes = replaceAll(code, "#include.*\n", m -> {
			p.includes.add(m.group());
			return "";
		});
		p.fileName = file.getName();
		p.code = maskedMinusIncludes;
		for (String key : unalterable.keySet()) {
			p.code = p.code.replaceAll(key, unalterable.get(key));
		}
		return p;
	}

	public static String handledUnalterable(String fullText, HashMap<String, String> unalterable) {
		String masked = replaceAll(fullText, "\\/\\/ <SingleHpp unalterable>(?s).*<\\/SingleHpp>", m -> {
			String id = UUID.randomUUID().toString();
			unalterable.put(id, m.group());
			return id;
		});
		return masked;
	}

	public static void create(String directory, String outputFile) throws IOException {
		FileUtils.writeFile(new File(outputFile), createTextFor(directory));
		SimpleLogger.event("Writing " + outputFile);
	}

	public static void main(String[] args) throws IOException {
		SimpleLogger.event("Starting...");
		create(".", args[0]);
	}
}
