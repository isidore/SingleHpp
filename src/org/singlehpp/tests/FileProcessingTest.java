package org.singlehpp.tests;

import static org.junit.Assert.assertEquals;

import java.io.File;
import java.util.HashMap;
import java.util.List;

import org.approvaltests.Approvals;
import org.junit.Test;
import org.singlehpp.Parts;
import org.singlehpp.SingleHpp;

public class FileProcessingTest {

	@Test
	public void testRemove() {

		assertEquals("  int i; \n  int j;", SingleHpp.removeComments("  int i; // this is the first value\n  int j;"));
		assertEquals("  int i;int j;",
				SingleHpp.removeComments("  int i;/* // \nthis is the first value\n   ** \n */int j;"));
	}

	@Test
	public void testGetFiles() throws Exception {
		Approvals.verifyAll("files", SingleHpp.getAllFiles("/Users/llewellyn/Github/ApprovalTests.cpp/ApprovalTests"));
	}

	@Test
	public void testGetSortedFiles() throws Exception {
		String directory = "/Users/llewellyn/Github/ApprovalTests.cpp/ApprovalTests";
		List<File> files = SingleHpp.getAllFiles(directory);
		List<Parts> parts = SingleHpp.getParts(files);
		Approvals.verifyAll("Sorted Files", parts, p -> p.fileName);
	}

	@Test
	public void testGetFileFromInclude() throws Exception {
		String[] includes = { "#include   \"..\\myfile.h\"", "#include   <string>" };
		Approvals.verifyAll("", includes, h -> h + " : " + Parts.getFileFromInclude(h));

	}

	@Test
	public void testUnalterable() throws Exception {
		String text = "// <SingleHpp unalterable>\n" + "#ifdef APPROVALS_CATCH\n" + "#define CATCH_CONFIG_MAIN\n"
				+ "#endif\n" + "\n" + "#include <Catch.hpp>\n" + "// </SingleHpp>";
		HashMap<String, String> unalterable = new HashMap<>();
		String result = SingleHpp.handledUnalterable(text, unalterable);
		Approvals.verifyAll("Unalterable", unalterable.values());

	}

	@Test
	public void testCreate() throws Exception {
		Approvals.verify(SingleHpp.createTextFor("/Users/llewellyn/Github/ApprovalTests.cpp/ApprovalTests"));
	}

	@Test
	public void demo() throws Exception {
		String outputFile = "/Users/llewellyn/Github/ApprovalTests.Cpp.StarterProject/lib/ApprovalTests.hpp";
		SingleHpp.create("/Users/llewellyn/Github/ApprovalTests.cpp/ApprovalTests", outputFile);
	}
}
