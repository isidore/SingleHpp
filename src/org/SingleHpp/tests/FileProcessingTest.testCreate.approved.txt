#include <string>
#include <fstream>
#include <cstring>
#include <exception>
#include <stdio.h>
#include <iostream>
#include <stdlib.h>
#include <unistd.h>
#include <vector>
#include <sys/stat.h>
#include <stack>
#include <sstream>
 // ******************** From: StringWriter.h
#ifndef STRINGWRITER_H
#define STRINGWRITER_H


class StringWriter
{
private:
    std::string s;
    std::string ext;

public:
    StringWriter( std::string contents, std::string fileExtensionWithDot = ".txt" )
        : s( contents ), ext( fileExtensionWithDot ) {}

    std::string GetFileExtension()
    {
        return ext;
    }

    void Write( std::string path )
    {
        std::ofstream out( path.c_str(), std::ofstream::out );
        this->Write( out );
        out.close();
    }

    void Write( std::ostream &out )
    {
        out << s << "\n";
    }


};
#endif

 // ******************** From: ApprovalException.h
#ifndef APPROVALEXCEPTION_H
#define APPROVALEXCEPTION_H


class ApprovalException : public std::exception
{
private:
    char *message;
protected:
    char *init( const char *msg )
    {
        size_t message_sz = std::strlen( msg );
        char *t = new char[message_sz + 1];
        std::strncpy( t, msg, message_sz + 1 );
        return t;
    };
public:
    ApprovalException( const char *msg ) : message( init( msg ) ) {}

    ApprovalException( const ApprovalException &a )
        : message( init( a.message ) )
    {
    }

    virtual ~ApprovalException() throw()
    {
        delete[] message;
    }

    virtual const char *what() const throw()
    {
        return message;
    }
};

class ApprovalMismatchException : public ApprovalException
{
private:
    char *format( const std::string &received, const std::string &approved )
    {
        size_t n = 2048;
        char s[n];
        int size = snprintf( s,
                             n,
                             "Failed Approval: \n"
                                     "Received does not match approved \n"
                                     "Received : \"%s\" \n"
                                     "Approved : \"%s\"",
                             received.c_str(),
                             approved.c_str() );
        char *t = new char[size + 1];
        std::strncpy( t, s, size + 1 );
        return t;
    }
public:
    ApprovalMismatchException( std::string received, std::string approved )
        : ApprovalException( format( received, approved ) )
    {
    }

    ApprovalMismatchException( const ApprovalMismatchException &a )
        : ApprovalException( a ) {}
};

class ApprovalMissingException : public ApprovalException
{
private:
    char *format( const std::string &file )
    {
        size_t n = 1024;
        char s[n];
        int size = snprintf( s,
                             n,
                             "Failed Approval: \n"
                                     "Approval File Not Found \n"
                                     "File: \"%s\"",
                             file.c_str() );
        char *t = new char[size + 1];
        std::strncpy( t, s, size + 1 );
        return t;
    }
public:
    ApprovalMissingException( std::string received, std::string approved )
        : ApprovalException( format( approved ) )
    {
    }

    ApprovalMissingException( const ApprovalMissingException &a )
        : ApprovalException( a )
    {
    }
};

#endif

 // ******************** From: CommandLauncher.h
#ifndef COMMANDLAUNCHER_H
#define COMMANDLAUNCHER_H


class CommandLauncher
{
public:
    virtual ~CommandLauncher() {}
    virtual bool Launch( std::vector<std::string> argv ) = 0;
};

class DoNothingLauncher : public CommandLauncher
{
private:
    std::string cmd;
public:
    bool working = true;
    bool Launch( std::vector<std::string> argv )
    {
        for ( std::vector<std::string>::iterator it = argv.begin();
                it != argv.end();
                ++it )
        {
            cmd += *it;
            cmd += " ";
        }
        return working;
    }

    std::string ReceivedCommand()
    {
        return cmd;
    }
};

class SystemLauncher : public CommandLauncher
{
public:
    bool exists(const std::string& command)
    {
        std::string which = "which " + command + " > /dev/null 2>&1";
        int result = system(which.c_str());
        return (result == 0);
    }

    bool Launch( std::vector<std::string> argv )
    {
        if (! exists(argv.front()))
        {
            return false;
        }

        char *ARGV[argv.size() + 1];
        int count = 0;

        for ( std::vector<std::string>::iterator it = argv.begin();
                it != argv.end();
                ++it )
        {
            char *cstr = new char [it->length() + 1];
            std::strcpy( cstr, it->c_str() );
            ARGV[count] = cstr;
            count++;
        }

        ARGV[argv.size()] = NULL;

        if ( fork() == 0 )
        {
            int i = execvp( ARGV[0], ARGV );

            if ( i < 0 )
            {
                std::cout << ARGV[0] << ": command not found" << std::endl;
                exit( 1 );
            }
        }

        for ( unsigned int i = 0; i < argv.size(); ++i )
        {
            delete ARGV[i];
        }

        return true;
    }
};

#endif

 // ******************** From: Macros.h
#ifndef CATCHPLAYGROUND_MARCOS_H
#define CATCHPLAYGROUND_MARCOS_H


#define STATIC(type, name, defaultValue) \
      static type &name(type *value = NULL) { \
static type *staticValue; \
if (value != NULL) { \
staticValue = value; \
} \
if (staticValue == NULL) \
{ \
 staticValue = defaultValue; \
} \
return *staticValue; \
} \



#endif 

 // ******************** From: FileUtils.h




#ifndef CATCHPLAYGROUND_FILEUTILS_H
#define CATCHPLAYGROUND_FILEUTILS_H


class FileUtils {
public:
    static bool fileExists(std::string path) {
        return fileSize(path) != -1;
    }

    static int fileSize(std::string path) {
        struct stat statbuf;
        int stat_ok = stat(path.c_str(), &statbuf);

        if (stat_ok == -1) {
            return -1;
        }

        return int(statbuf.st_size);
    }

    static void ensureFileExists(std::string fullFilePath) {
        if (!fileExists(fullFilePath)) {
            StringWriter s("", "");
            s.Write(fullFilePath);
        }
    }
};

#endif 

 // ******************** From: ApprovalNamer.h
#ifndef APPROVALNAMER_H
#define APPROVALNAMER_H


using std::string;

class TestName {
public:
    string fileName;
    string testCase;
    std::vector<string> sections;
};

class ApprovalNamer {
private:
public:
    ApprovalNamer() {
    }

    string getTestName() {
        std::stringstream ext;
        auto test = currentTest();
        for (int i = 0; i < test.sections.size(); i++) {
            if (0 < i) {
                ext << ".";
            }
            ext << test.sections[i];

        }

        return ext.str();
    }

    string getFileName() {
        auto file = currentTest().fileName;
        auto start = file.rfind("/") + 1;
        auto end = file.rfind(".");
        return file.substr(start, end - start);
    }

    string getDirectory() {
        auto file = currentTest().fileName;
        auto end = file.rfind("/") + 1;
        return file.substr(0, end);
    }

    STATIC(TestName, currentTest, NULL)

    string getApprovedFile(string extentionWithDot) {

        return getFullFileName(".approved", extentionWithDot);
    }

    string getReceivedFile(string extentionWithDot) {

        return getFullFileName(".received", extentionWithDot);
    }

    string getFullFileName(string approved, string extentionWithDot) {
        std::stringstream ext;
        ext << getDirectory() << getFileName() << "." << getTestName() << approved << extentionWithDot;
        return ext.str();
    }
};

#endif

 // ******************** From: Reporter.h
#ifndef REPORTER_H
#define REPORTER_H


class Reporter {
public:
    virtual bool Report(std::string received, std::string approved) const = 0;
};

class CommandReporter : public Reporter {
private:
    std::string cmd;
    CommandLauncher *l;

protected:
    CommandReporter(std::string command, CommandLauncher *launcher)
            : cmd(command), l(launcher) {
    }

public:
    bool Report(std::string received, std::string approved) const override {
        FileUtils::ensureFileExists(approved);
        std::vector<std::string> fullCommand;
        fullCommand.push_back(cmd);
        fullCommand.push_back(received);
        fullCommand.push_back(approved);
        return l->Launch(fullCommand);
    }
};

class GenericDiffReporter : public CommandReporter {
private:
    SystemLauncher launcher;

public:
    GenericDiffReporter(const std::string& program) : CommandReporter(program, &launcher) {};
};
class MeldReporter : public GenericDiffReporter {
public:
    MeldReporter() : GenericDiffReporter("meld") {};
};

class DiffMergeReporter : public GenericDiffReporter {
public:
    DiffMergeReporter() : GenericDiffReporter("/Applications/DiffMerge.app/Contents/MacOS/DiffMerge") {};
};

class TestReporter : public CommandReporter {
public:
    DoNothingLauncher launcher;

    TestReporter(bool working = true) : CommandReporter("fake", &launcher) {
        launcher.working = working;
    };
};

class ReporterFactory {
private:
    ReporterFactory() {}

    ~ReporterFactory() {}

public:
    STATIC(Reporter, currentReporter, new MeldReporter())

    static Reporter &getCurrentReporter() {
        return currentReporter();
    }



    template<typename ReporterType>
    static ReporterType &UseReporter() {
        delete &getCurrentReporter();
        ReporterType *t = new ReporterType();
        return currentReporter(t);
    }
};

#endif

 // ******************** From: FileApprover.h
#ifndef FILEAPPROVER_H
#define FILEAPPROVER_H


class FileApprover {
public:
    FileApprover() {};

    ~FileApprover() {};

    static ApprovalException *verify(std::string receivedPath,
                                     std::string approvedPath) {
        int asize = FileUtils::fileSize(approvedPath);

        if (-1 == asize) {
            return new ApprovalMissingException(receivedPath, approvedPath);
        }

        int rsize = FileUtils::fileSize(receivedPath);

        if (-1 == rsize) {
            return new ApprovalMissingException(approvedPath, receivedPath);
        }

        if (asize != rsize) {
            return new ApprovalMismatchException(receivedPath, approvedPath);
        }

        std::ifstream astream(approvedPath.c_str(),
                              std::ios::binary | std::ifstream::in);
        std::ifstream rstream(receivedPath.c_str(),
                              std::ios::binary | std::ifstream::in);

        while (astream.good() && rstream.good()) {
            int a = astream.get();
            int r = rstream.get();

            if (a != r) {
                return new ApprovalMismatchException(receivedPath, approvedPath);
            }
        }

        remove(receivedPath.c_str());
        return NULL;
    }


    static void verify(ApprovalNamer n, StringWriter s, const Reporter& r) {
        std::string approvedPath = n.getApprovedFile(s.GetFileExtension());
        std::string receivedPath = n.getReceivedFile(s.GetFileExtension());
        s.Write(receivedPath);
        ApprovalException *ae = verify(receivedPath, approvedPath);

        if (ae != NULL) {
            r.Report(receivedPath, approvedPath);
            ApprovalException e(*ae);
            delete ae;
            throw e;
        } else {
            std::cout << "Removing " << receivedPath;
            remove(receivedPath.c_str());
        }
    }


};

#endif

 // ******************** From: FirstWorkingReporter.h
#ifndef CATCHPLAYGROUND_FIRSTWORKINGREPORTER_H
#define CATCHPLAYGROUND_FIRSTWORKINGREPORTER_H


class FirstWorkingReporter : public Reporter
{
private:
    std::vector<Reporter*> reporters;
public:
    FirstWorkingReporter(std::vector<Reporter*> theReporters) : reporters(theReporters)
    {
    }

    bool Report(std::string received, std::string approved) const override
    {
        for(auto r : reporters)
        {
            if (r->Report(received, approved))
            {
                return true;
            }
        }
        return false;
    }
};

#endif 

 // ******************** From: DiffReporter.h
#ifndef CATCHPLAYGROUND_DIFFREPORTER_H
#define CATCHPLAYGROUND_DIFFREPORTER_H


class MacDiffReporter : public FirstWorkingReporter
{
public:
    MacDiffReporter() : FirstWorkingReporter({new DiffMergeReporter()})
    {
    }
};

class LinuxDiffReporter : public FirstWorkingReporter
{
public:
    LinuxDiffReporter() : FirstWorkingReporter({new MeldReporter()})
    {
    }
};

class WindowsDiffReporter : public FirstWorkingReporter
{
public:
    WindowsDiffReporter() : FirstWorkingReporter({})
    {
    }
};

class DiffReporter : public FirstWorkingReporter
{
public:
    DiffReporter() : FirstWorkingReporter({new MacDiffReporter(), new LinuxDiffReporter(), new WindowsDiffReporter()})
    {
    }

};

#endif 

 // ******************** From: Approvals.h
#ifndef APPROVALS_H
#define APPROVALS_H


class Approvals
{
private:
    Approvals() {}
    ~Approvals() {}
public:
    static void verify( std::string contents,const Reporter& reporter = DiffReporter() )
    {
        StringWriter writer( contents );
        ApprovalNamer namer;
        FileApprover::verify(namer, writer, reporter);
    }
};

#endif

 // ******************** From: ApprovalTests.h




#ifndef CATCHPLAYGROUND_APPROVALTESTS_H_H
#define CATCHPLAYGROUND_APPROVALTESTS_H_H


#endif 

 // ******************** From: Catch2Approvals.h

#ifndef CATCHPLAYGROUND_CATCH2APPROVALS_H_H
#define CATCHPLAYGROUND_CATCH2APPROVALS_H_H

// <SingleHpp unalterable>
#ifdef APPROVALS_CATCH
#define CATCH_CONFIG_MAIN
#endif

#include <Catch.hpp>
// </SingleHpp>

using std::string;

#ifdef APPROVALS_CATCH

struct Catch2ApprovalListener : Catch::TestEventListenerBase {
    using TestEventListenerBase::TestEventListenerBase; 
    TestName currentTest;

    virtual void testCaseStarting(Catch::TestCaseInfo const &testInfo) override {
        
        currentTest.fileName = testInfo.lineInfo.file;
        currentTest.testCase = testInfo.name;
        ApprovalNamer::currentTest(&currentTest);
    }

    virtual void testCaseEnded(Catch::TestCaseStats const &testCaseStats) override {
        while (!currentTest.sections.empty()) {
            currentTest.sections.pop_back();
        }
    }

    virtual void sectionStarting(Catch::SectionInfo const &sectionInfo) override {
        currentTest.sections.push_back(sectionInfo.name);
    }

    virtual void sectionEnded(Catch::SectionStats const &sectionStats) override {
        currentTest.sections.pop_back();
    }
};

CATCH_REGISTER_LISTENER(Catch2ApprovalListener)

#endif
#endif 

